Setting up dynamic language switch

### Discussed in https://github.com/Tinkoff/taiga-ui/discussions/2990

<div type='discussions-op-text'>

<sup>Originally posted by **loxy** October 31, 2022</sup>
Hi,

I cannot setup dynamic language switching like described here: https://taiga-ui.dev/i18n/Dynamic_loader

I copied the dynamic loading to our app module. But when changing via `this.switcher.setLanguage('english')` I get:

```sh
ERROR Error: Cannot find module './english'
    at .*$ chunkName: i18n-lazy- namespace object:1366:1
```

In this chunk I've found several entries like this (here only english):

```json5
// ...
"./english/addon-commerce.d": [
	4699,
	7,
	"i18n-lazy-42"
],
"./english/addon-commerce.d.ts": [
	4699,
	7,
	"i18n-lazy-42"
],
"./english/addon-editor.d": [
	1870,
	7,
	"i18n-lazy-44"
],
"./english/addon-editor.d.ts": [
	1870,
	7,
	"i18n-lazy-44"
],
"./english/addon-preview.d": [
	3976,
	7,
	"i18n-lazy-46"
],
"./english/addon-preview.d.ts": [
	3976,
	7,
	"i18n-lazy-46"
],
"./english/addon-table.d": [
	6475,
	7,
	"i18n-lazy-48"
],
"./english/addon-table.d.ts": [
	6475,
	7,
	"i18n-lazy-48"
],
"./english/core.d": [
	9147,
	7,
	"i18n-lazy-50"
],
"./english/core.d.ts": [
	9147,
	7,
	"i18n-lazy-50"
],
"./english/countries.d": [
	4312,
	7,
	"i18n-lazy-52"
],
"./english/countries.d.ts": [
	4312,
	7,
	"i18n-lazy-52"
],
"./english/english.d": [
	5875,
	7,
	"i18n-lazy-54"
],
"./english/english.d.ts": [
	5875,
	7,
	"i18n-lazy-54"
],
"./english/index.d": [
	7658,
	7,
	"i18n-lazy-56"
],
"./english/index.d.ts": [
	7658,
	7,
	"i18n-lazy-56"
],
"./english/kit.d": [
	6678,
	7,
	"i18n-lazy-58"
],
"./english/kit.d.ts": [
	6678,
	7,
	"i18n-lazy-58"
],
"./english/package.json": [
	6223,
	3,
	"i18n-lazy-60"
],
"./english/taiga-ui-i18n-languages-english.d": [
	5794,
	7,
	"i18n-lazy-61"
],
"./english/taiga-ui-i18n-languages-english.d.ts": [
	5794,
	7,
	"i18n-lazy-61"
],
// ...
```

Changing the dynamic loading to something like this:

```ts
@NgModule({
  providers: [
    tuiLanguageSwitcher(
      async (language: TuiLanguageName): Promise<unknown> =>
        import(
          /* webpackMode: "lazy" */
          /* webpackChunkName: "i18n-lazy-" */
          `@taiga-ui/i18n/languages/${language}/index.d`
          // also you can override the paths to your i18n language files
        ),
    ),
  ],
})
export class AppModule {}
```
solves this webpack issue but leads to 

```sh
2i18n-lazy-2.js:1 ERROR TypeError: Cannot read properties of undefined (reading 'dateTexts')
    at taiga-ui-i18n-tools.js:7:157
    at map.js:7:37
    at OperatorSubscriber._next (OperatorSubscriber.js:13:21)
    at OperatorSubscriber.next (Subscriber.js:31:18)
    at subscribe.innerSubscriber (switchMap.js:14:144)
    at OperatorSubscriber._next (OperatorSubscriber.js:13:21)
    at OperatorSubscriber.next (Subscriber.js:31:18)
    at map.js:7:24
    at OperatorSubscriber._next (OperatorSubscriber.js:13:21)
    at OperatorSubscriber.next (Subscriber.js:31:18)
    ...
```

Can you elaborate more on how this should work or being set up? I'm lost and didn't found any hints in code base...

I'm on Taiga UI 3.8.0 and Angular 14

Best regards,
Kersten</div>

--- a/projects/demo/src/modules/customization/i18n/app.esbuild.module.md
+++ b/projects/demo/src/modules/customization/i18n/app.esbuild.module.md
@@ -1,0 +1,48 @@
+```ts
+import {TuiLanguageName} from '@taiga-ui/i18n/interfaces';
+import {tuiLanguageSwitcher} from '@taiga-ui/i18n/switch';
+
+@NgModule({
+  imports: [
+    // ...
+  ],
+  providers: [
+    // ...
+    tuiLanguageSwitcher(async (language: TuiLanguageName): Promise<unknown> => {
+      switch (language) {
+        case `chinese`:
+          return import(`@taiga-ui/i18n/languages/chinese`);
+        case `dutch`:
+          return import(`@taiga-ui/i18n/languages/dutch`);
+        case `french`:
+          return import(`@taiga-ui/i18n/languages/french`);
+        case `german`:
+          return import(`@taiga-ui/i18n/languages/german`);
+        case `italian`:
+          return import(`@taiga-ui/i18n/languages/italian`);
+        case `polish`:
+          return import(`@taiga-ui/i18n/languages/polish`);
+        case `portuguese`:
+          return import(`@taiga-ui/i18n/languages/portuguese`);
+        case `russian`:
+          return import(`@taiga-ui/i18n/languages/russian`);
+        case `spanish`:
+          return import(`@taiga-ui/i18n/languages/spanish`);
+        case `turkish`:
+          return import(`@taiga-ui/i18n/languages/turkish`);
+        case `ukrainian`:
+          return import(`@taiga-ui/i18n/languages/ukrainian`);
+        case `vietnamese`:
+          return import(`@taiga-ui/i18n/languages/vietnamese`);
+
+        case `serbian`: // custom language for example
+          return import(`@my-scope/serbian-language`);
+
+        default:
+          return import(`@taiga-ui/i18n/languages/english`);
+      }
+    }),
+  ],
+})
+export class AppModule {}
+```

--- a/projects/demo/src/modules/customization/i18n/i18n.component.ts
+++ b/projects/demo/src/modules/customization/i18n/i18n.component.ts
@@ -19,6 +19,7 @@ export class I18nComponent {
     ).pipe(map(readme => readme.default.split('Supported languages:')[1]));
 
     exampleModule = import('./app.module.md?raw');
+    exampleEsbuildModule = import('./app.esbuild.module.md?raw');
 
     example1: TuiDocExample = {
         'language-switcher.component.html': import(

--- a/projects/demo/src/modules/customization/i18n/i18n.template.html
+++ b/projects/demo/src/modules/customization/i18n/i18n.template.html
@@ -15,5 +15,22 @@
         >
             <tui-language-switcher>Language</tui-language-switcher>
         </tui-doc-example>
+
+        <h2 class="tui-space_top-10">CLI's esbuild-based build system</h2>
+        <p>
+            Dynamic import expressions that do not contain static values will be kept in their original form and not
+            processed at build time. This is a limitation of the underlying bundler but is
+            <a
+                href="https://github.com/evanw/esbuild/pull/2508"
+                tuiLink
+            >
+                planned
+            </a>
+            to be implemented in the future. In many cases, application code can be made to work by changing the import
+            expressions into static strings with some form of conditional statement such as an if or switch for the
+            known potential files.
+        </p>
+
+        <tui-doc-code [code]="exampleEsbuildModule"></tui-doc-code>
     </ng-template>
 </tui-doc-page>

